### PR TITLE
Introduced new tree navigation

### PR DESCRIPTION
**Pull request by honza-kasik**
Introduced new, more flexible and organized tree navigation based only on Jquery selectors. this was created to test generated subsystems.

### CODE REVIEW

File: common/src/main/java/org/jboss/hal/testsuite/mbui/MBUITreeNavigation.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        return this;
+    }
+
+    /**
+     * Navigates to target item step by step
+     * @return target item
+     */
+    public MBUITreeGroup navigateToTreeItem() {
+        MBUITreeGroup current = new MBUITreeGroup(root.findElement(ByJQuery.selector("> div:nth-child(2)")));
+        logger.info(current.getRoot().getAttribute("innerHTML"));
```
Comment (zebrik) [inline common/src/main/java/org/jboss/hal/testsuite/mbui/MBUITreeNavigation.java]: Might be desirable to add some context prefix to the log output.

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/SimpleWebserviceEndpointTestCase.java [hunk 2]
```diff
@@ -159,10 +161,13 @@ public void callServiceAndGetMetrics() throws Exception {
 
         wsePage.navigateInDeploymentsMenu();
 
-        mbuiNavigation.displayMBuiSubTree("subsystem");
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/SimpleWebserviceEndpointTestCase.java:162]: I suggest to mark `org.jboss.hal.testsuite.mbui.MbuiNavigation.MbuiNavigation(WebDriver)` deprecated.

**Review by zebrik — APPROVED** [review]
Generally LGTM.

**Comment by honza-kasik** [comment]
This can be now merged. @rhatlapa @zebrik Can you please review it if you have some time?

File: common/src/main/java/org/jboss/hal/testsuite/page/config/TreeNavigationPage.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import org.jboss.hal.testsuite.treefinder.TreeNavigation;
+
+/**
+ * Page which contains tree navigation in left column and form table with values in right column
+ */
+public class TreeNavigationPage extends ConfigPage {
+
+    public TreeNavigation getTreeNavigation() {
+        return Graphene.createPageFragment(TreeNavigation.class, getContentRoot());
+    }
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/page/config/TreeNavigationPage.java:16]: I suggest renaming to `treeNavigation`. Note this is no getter and there is no reason for the longer name containing the `get` keyword

File: common/src/main/java/org/jboss/hal/testsuite/page/config/TreeNavigationPage.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+    public FormItemTableFragment getFromItemTable() {
+        return Graphene.createPageFragment(FormItemTableFragment.class,
+                getContentRoot().findElement(ByJQuery.selector("table.form-item-table")));
+    }
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/page/config/TreeNavigationPage.java:21]: I suggest renaming to `formItemTable`

File: common/src/main/java/org/jboss/hal/testsuite/page/runtime/WebServiceEndpointsPage.java [hunk 5]
```diff
@@ -50,4 +54,14 @@ public void navigate() {
...
     }
+
+    public TreeNavigation getTreeNavigation() {
+        return Graphene.createPageFragment(TreeNavigation.class, getContentRoot());
+    }
+
+    public FormItemTableFragment getFromItemTable() {
+        return Graphene.createPageFragment(FormItemTableFragment.class,
+                getContentRoot().findElement(ByJQuery.selector("table.form-item-table")));
+    }
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/page/runtime/WebServiceEndpointsPage.java]: The same notes for naming as for `TreeNavigationPage`. Also why this page requires the same code as `TreeNavigationPage` class?

File: common/src/main/java/org/jboss/hal/testsuite/treefinder/TreeNavigation.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        }
+        steps.add(label);
+        return this;
+    }
+
+    /**
+     * Navigates to target item step by step
+     * @return target item
+     */
+    public TreeNavigationGroup navigateToTreeItem() {
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/treefinder/TreeNavigation.java:53]: Why not just name it `navigate` ?

File: common/src/main/java/org/jboss/hal/testsuite/treefinder/TreeNavigationGroup.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+    private static final String TREE_ITEM_CLASS_NAME = "gwt-TreeItem";
+
+    private static final By DIRECT_CHILDREN_SELECTOR = ByJQuery.selector("> div > div");
+
+    private final WebElement root;
+
+    private TreeNavigationGroup() {
+        throw new UnsupportedOperationException("It is not allowed to use this constructor!");
+    }
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/treefinder/TreeNavigationGroup.java]: When it is not supported, why you have it?

File: common/src/main/java/org/jboss/hal/testsuite/treefinder/TreeNavigationGroup.java [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    public String getLabel() {
+        return root.findElement(ByJQuery.selector("div." + TREE_ITEM_CLASS_NAME)).getText().trim();
+    }
+
+    /**
+     * Return direct child of current root tree item by its label
+     * @param label by which the item will be looked for
+     * @return this
+     */
+    TreeNavigationGroup getDirectChildByLabel(String label) {
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/treefinder/TreeNavigationGroup.java:66]: You are not returning this, you are returning the child => fix the javadoc

File: common/src/main/java/org/jboss/hal/testsuite/treefinder/TreeNavigationGroup.java [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        for (TreeNavigationGroup child : getDirectChildren()) {
+            if (child.getLabel().equals(label)) {
+                return child;
+            }
+        }
+        //TODO: Is it good thing to throw an unchecked exception here? Should not be the user of API informed?
+        throw new NotFoundException("Not found child with label '" + label + "'!");
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/treefinder/TreeNavigationGroup.java:72]: You can at least mention it in JavaDoc

File: common/src/main/java/org/jboss/hal/testsuite/treefinder/TreeNavigationGroup.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    TreeNavigationGroup openSubTreeIfNotOpen() {
+        if (!root.findElement(DIRECT_CHILDREN_SELECTOR).isDisplayed()) {
+            clickOpenTreeToggleButton();
+        }
+        return this;
+    }
+
+    /**
+     * Find if this item has a subtree to open
+     * @return true if this item has subtree
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/treefinder/TreeNavigationGroup.java]: when it returns false? ;-). I suggest to add `, false otherwise`

File: common/src/main/java/org/jboss/hal/testsuite/treefinder/TreeNavigationGroup.java [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+     */
+    public boolean hasSubtree() {
+        //Just check if this root contains table with arrow and label
+        return root.findElements(ByJQuery.selector("> table")).size() > 0;
+    }
+
+    /**
+     * Find if this item has children
+     * @return true if this item has children
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/treefinder/TreeNavigationGroup.java]: `false otherwise`

**Review by rhatlapa — APPROVED** [review]
Generally LGTM, several suggestions to naming + to javadoc.

File: common/src/main/java/org/jboss/hal/testsuite/treefinder/TreeNavigation.java (hunk 6 above)
Comment (honza-kasik) [reply common/src/main/java/org/jboss/hal/testsuite/treefinder/TreeNavigation.java:53]: I wanted to highlight the fact that this method returns something. WDYT?

File: common/src/main/java/org/jboss/hal/testsuite/treefinder/TreeNavigationGroup.java (hunk 7 above)
Comment (honza-kasik) [reply common/src/main/java/org/jboss/hal/testsuite/treefinder/TreeNavigationGroup.java]: To override implicit constructor and disallow instantiating `TreeNavigationGroup` without arguments.

Comment (rhatlapa) [reply common/src/main/java/org/jboss/hal/testsuite/treefinder/TreeNavigationGroup.java]: The implicit is used only if there is no constructor defined in the class. In this case, you have constructor with one param => implicit constructor without params isn't available.

File: common/src/main/java/org/jboss/hal/testsuite/page/runtime/WebServiceEndpointsPage.java (hunk 5 above)
Comment (honza-kasik) [reply common/src/main/java/org/jboss/hal/testsuite/page/runtime/WebServiceEndpointsPage.java]: Well, `WebServiceEndpointsPage` is already extending `MetricsPage`. AFAIK not every `MetricsPage` is equipped with tree navigation. There are several other solutions available:
1. Let `MetricsPage` extend `TreeNavigationPage` despite the fact that is-a rule will be corrupted
2. Make methods in `TreeNavigationPage` static and call them from wrappers in `WebServicesEndpointsPage` to avoid some of the redundancy
3. Keep the state as it is in this pull request
4. Some other?

WDYT?

File: common/src/main/java/org/jboss/hal/testsuite/treefinder/TreeNavigationGroup.java (hunk 7 above)
Comment (honza-kasik) [reply common/src/main/java/org/jboss/hal/testsuite/treefinder/TreeNavigationGroup.java]: My bad, thanks!

**Comment by honza-kasik** [comment]
Ready to be merged!

File: common/src/main/java/org/jboss/hal/testsuite/page/runtime/WebServiceEndpointsPage.java (hunk 5 above)
Comment (zebrik) [reply common/src/main/java/org/jboss/hal/testsuite/page/runtime/WebServiceEndpointsPage.java]: The redundancy of code shows you that you are using wrong page. As soon as you will need to call `treeNavigation()` or `formItemTable()` it  will mean that you will have already opened `TreeNavigationPage` and you should use it instead of `WebServiceEndpointsPage`. 
So I suggest to remove that methods from `WebServiceEndpointsPage`.

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/SimpleWebserviceEndpointTestCase.java [hunk 12]
```diff
@@ -159,14 +155,17 @@ public void callServiceAndGetMetrics() throws Exception {
 
         wsePage.navigateInDeploymentsMenu();
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/SimpleWebserviceEndpointTestCase.java:161]: Is it possible to 
1. move `WebServiceEndpointsPage.navigateInDeploymentsMenu()` to `DeploymentPage` ?
2. merge `DeploymentPage` with `DomainDeploymentPage` ?
Might be in another pull request.

**Review by zebrik — COMMENTED** [review]
Generally LGTM. Just suggestions related to removing unrelated code from `WebServiceEndpointsPage`:

The redundancy of code in `WebServiceEndpointsPage` shows you that you are using wrong page. As soon as you will need to call `treeNavigation()` or `formItemTable()` it will mean that you will have already 'opened' `TreeNavigationPage` and you should use it instead of `WebServiceEndpointsPage`.
So I suggest to remove that methods from `WebServiceEndpointsPage`.

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/SimpleWebserviceEndpointTestCase.java (hunk 12 above)
Comment (honza-kasik) [reply basic/src/test/java/org/jboss/hal/testsuite/test/runtime/SimpleWebserviceEndpointTestCase.java:161]: I am going to split one used page to several, better do it now, than forget about it...

**Comment by honza-kasik** [comment]
Ready to be merged - problem with deployment pages will be solved in next pull request.

**Comment by zebrik** [comment]
LGTM now. Thank you. 
Please do WebServiceEndpointsPage, DeploymentPage, DomainDeploymentPage cleanup in another PR. Thank you in advance!
